### PR TITLE
PS-4790: Improve user statistics accuracy (8.0)

### PR DESCRIPTION
**Pull request by inikep**
Measure times for `dispatch_command` instead of mysql_parse, mysqld_stmt_reset, mysqld_stmt_fetch, mysqld_stmt_prepare, mysqld_stmt_execute.

### CODE REVIEW

File: sql/sql_parse.cc [hunk 1]
```diff
@@ -1448,6 +1448,32 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
   if (!(server_command_flags[command] & CF_SKIP_QUESTIONS))
     thd->status_var.questions++;
 
+  /* Declare userstat variables */
+  int start_time_error = 0;
+  int end_time_error = 0;
+  struct timeval start_time, end_time;
+  double start_usecs = 0;
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: 0.0

File: sql/sql_parse.cc [hunk 2]
```diff
@@ -1448,6 +1448,32 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+  double end_usecs = 0;
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: 0.0

File: sql/sql_parse.cc [hunk 3]
```diff
@@ -1448,6 +1448,32 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+  int cputime_error = 0;
+#ifdef HAVE_CLOCK_GETTIME
+  struct timespec tp;
+#endif
+  double start_cpu_nsecs = 0;
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: 0.0

File: sql/sql_parse.cc [hunk 4]
```diff
@@ -1448,6 +1448,32 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+
+  if (unlikely(opt_userstat)) {
+#ifdef HAVE_CLOCK_GETTIME
+    /* Get start cputime */
+    if (!(cputime_error = clock_gettime(CLOCK_THREAD_CPUTIME_ID, &tp)))
+      start_cpu_nsecs = tp.tv_sec * 1000000000.0 + tp.tv_nsec;
+#endif
+
+    /* Gets the start time, in order to measure how long this command takes. */
+    if (!(start_time_error = gettimeofday(&start_time, NULL))) {
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: nullptr

File: sql/sql_parse.cc [hunk 5]
```diff
@@ -1993,6 +2019,46 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
   DBUG_ASSERT(thd->open_tables == NULL ||
               (thd->locked_tables_mode == LTM_LOCK_TABLES));
 
+  /* Update user statistics */
+  if (unlikely(opt_userstat)) {
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: There is a race condition in that opt_userstat can change its value between the two sections. So in the second section I'd only check for start_* timestamp initialization

File: sql/sql_parse.cc [hunk 6]
```diff
@@ -1993,6 +2019,46 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+    /* Gets the end time. */
+    if (!(end_time_error = gettimeofday(&end_time, NULL))) {
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: nullptr

File: sql/sql_parse.cc [hunk 7]
```diff
@@ -1993,6 +2019,46 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+      end_usecs = end_time.tv_sec * 1000000.0 + end_time.tv_usec;
+    }
+
+    /* Calculates the difference between the end and start times. */
+    if (start_usecs && end_usecs >= start_usecs && !start_time_error &&
+        !end_time_error) {
+      thd->busy_time = (end_usecs - start_usecs) / 1000000;
+      /* In case there are bad values, 2629743 is the #seconds in a month. */
+      if (thd->busy_time > 2629743) {
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: unlikely

File: sql/sql_parse.cc [hunk 8]
```diff
@@ -1993,6 +2019,46 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
...
+#ifdef HAVE_CLOCK_GETTIME
+    /* Get end cputime */
+    if (!cputime_error &&
+        !(cputime_error = clock_gettime(CLOCK_THREAD_CPUTIME_ID, &tp)))
+      end_cpu_nsecs = tp.tv_sec * 1000000000.0 + tp.tv_nsec;
+#endif
+    if (start_cpu_nsecs && !cputime_error) {
+      thd->cpu_time = (end_cpu_nsecs - start_cpu_nsecs) / 1000000000;
+      /* In case there are bad values, 2629743 is the #seconds in a month. */
+      if (thd->cpu_time > 2629743) {
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: unlikely

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Nice cleanup. Minor comments

**Comment by inikep** [comment]
Added `userstat.cc` and `userstat.h` with the following functions:
```
void userstat_start_timer(double &start_busy_usecs, double &start_cpu_nsecs);
void userstat_finish_timer(double &start_busy_usecs, double &start_cpu_nsecs,
                           double &busy_sec, double &cpu_sec);
```

**Comment by inikep** [comment]
use pointer for output arg:
```
void userstat_start_timer(double *start_busy_usecs, double *start_cpu_nsecs);
void userstat_finish_timer(double &start_busy_usecs, double &start_cpu_nsecs,
                           double *busy_sec, double *cpu_sec);
```

File: sql/sql_parse.cc [hunk 9]
```diff
@@ -1993,6 +2000,15 @@ bool dispatch_command(THD *thd, const COM_DATA *com_data,
   DBUG_ASSERT(thd->open_tables == NULL ||
               (thd->locked_tables_mode == LTM_LOCK_TABLES));
 
+  /* Update user statistics only if at least one timer was initialized */
+  if (unlikely(start_busy_usecs > 0.0 || start_cpu_nsecs > 0.0)) {
+    userstat_finish_timer(start_busy_usecs, start_cpu_nsecs, &thd->busy_time,
+                          &thd->cpu_time);
+    /* Updates THD stats and the global user stats. */
+    thd->update_stats(true);
+    update_global_user_stats(thd, true, time(NULL));
```
Comment (laurynas-biveinis) [inline sql/sql_parse.cc]: nullptr

File: sql/userstat.cc [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+  userstat_start_timer() and this function.
+
+  @param  start_busy_usecs  Start value of busy_time or 0.0 in case of error
+  @param  start_cpu_nsecs   Start value of cpu_time or 0.0 in case of error
+  @return busy_sec          Interval time for busy_time in seconds or 0.0 in
+                            case of error
+  @return cpu_sec           Interval time for cpu_time in seconds or 0.0 in case
+                            of error
+*/
+void userstat_finish_timer(double &start_busy_usecs, double &start_cpu_nsecs,
```
Comment (laurynas-biveinis) [inline sql/userstat.cc]: Pass by value start args

File: sql/userstat.cc [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+#include "userstat.h"
+
+/**
+  Get start timers for cpu_time and busy_time.
+
+  @return start_busy_usecs  Start value of busy_time or 0.0 in case of error
+  @return start_cpu_nsecs   Start value of cpu_time or 0.0 in case of error
+*/
+void userstat_start_timer(double *start_busy_usecs, double *start_cpu_nsecs) {
+  struct timeval start_time;
```
Comment (laurynas-biveinis) [inline sql/userstat.cc]: Declare at use

File: sql/userstat.cc [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+  @return busy_sec          Interval time for busy_time in seconds or 0.0 in
+                            case of error
+  @return cpu_sec           Interval time for cpu_time in seconds or 0.0 in case
+                            of error
+*/
+void userstat_finish_timer(double &start_busy_usecs, double &start_cpu_nsecs,
+                           double *busy_sec, double *cpu_sec) {
+  struct timeval end_time;
+  double end_busy_usecs = 0.0;
+  double end_cpu_nsecs = 0.0;
```
Comment (laurynas-biveinis) [inline sql/userstat.cc]: Declare at use

File: sql/userstat.h [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */
+
+#ifndef USERSTAT_INCLUDED
+#define USERSTAT_INCLUDED
+
+void userstat_start_timer(double *start_busy_usecs, double *start_cpu_nsecs);
```
Comment (laurynas-biveinis) [inline sql/userstat.h]: both are noexcept

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Minor comments

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**